File: backend/app/models/application.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Application(Base):
    __tablename__ = "applications"
    __table_args__ = (UniqueConstraint("user_id", "job_id", name="uq_user_job_application"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    job_id: Mapped[int] = mapped_column(ForeignKey("jobs.id", ondelete="CASCADE"), nullable=False)
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="applied")
    cover_letter: Mapped[str | None] = mapped_column(Text, nullable=True)
    applied_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    user = relationship("User", back_populates="applications")
    job = relationship("Job", back_populates="applications")

File: frontend/lib/models/application.dart
import 'job.dart';

class Application {
  final int id;
  final int userId;
  final int jobId;
  final String status;
  final String? coverLetter;
  final DateTime? appliedAt;
  final Job? job;

  Application({
    required this.id,
    required this.userId,
    required this.jobId,
    required this.status,
    this.coverLetter,
    this.appliedAt,
    this.job,
  });

  factory Application.fromJson(Map<String, dynamic> json) {
    return Application(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      jobId: json['job_id'] as int,
      status: json['status'] as String,
      coverLetter: json['cover_letter'] as String?,
      appliedAt: json['applied_at'] != null ? DateTime.tryParse(json['applied_at']) : null,
      job: json['job'] != null ? Job.fromJson(json['job']) : null,
    );
  }
}
